examples/poyo/train.py

```diff
@@ -0,0 +1,362 @@
+        ),
+        LearningRateMonitor(logging_interval="step"),
+        tbrain_callbacks.MemInfo(),
+        tbrain_callbacks.EpochTimeLogger(),
+        tbrain_callbacks.ModelWeightStatsLogger(),
+    ]
+
+    trainer = L.Trainer(
+        logger=wandb_logger,
+        default_root_dir=cfg.log_dir,
+        check_val_every_n_epoch=cfg.eval_epochs,
+        max_epochs=cfg.epochs,
+        log_every_n_steps=1,
+        strategy=(
+            "ddp_find_unused_parameters_true" if torch.cuda.is_available() else "auto"
+        ),
+        callbacks=callbacks,
+        precision=cfg.precision,
+        accelerator="gpu" if torch.cuda.is_available() else "cpu",
+        devices=cfg.gpus,
+        num_nodes=cfg.nodes,
+        limit_val_batches=None,  # Ensure no limit on validation batches
+        num_sanity_val_steps=cfg.num_sanity_val_steps,
+    )
+
+    # Train
+    trainer.fit(wrapper, data_module, ckpt_path=cfg.ckpt_path)
+
+    # Test
+    trainer.test(wrapper, data_module)
+
+
+class POYOTrainWrapper(L.LightningModule):
+    def __init__(
+        self,
+        cfg: DictConfig,
+        model: nn.Module,
+        modality_spec: ModalitySpec,
+    ):
+        super().__init__()
+
+        self.cfg = cfg
+        self.model = model
+        self.modality_spec = modality_spec
+        self.save_hyperparameters(OmegaConf.to_container(cfg))
+
+    def configure_optimizers(self):
+        max_lr = self.cfg.optim.base_lr * self.cfg.batch_size  # linear scaling rule
+
+        optimizer = Lamb(
+            self.model.parameters(),
+            lr=max_lr,
+            weight_decay=self.cfg.optim.weight_decay,
+        )
+
+        scheduler = torch.optim.lr_scheduler.OneCycleLR(
+            optimizer,
+            max_lr=max_lr,
+            total_steps=self.trainer.estimated_stepping_batches,
+            pct_start=self.cfg.optim.lr_decay_start,
+            anneal_strategy="cos",
+            div_factor=1,
+        )
+
+        return {
+            "optimizer": optimizer,
+            "lr_scheduler": {
+                "scheduler": scheduler,
+                "interval": "step",
+            },
+        }
+
+    def training_step(self, batch, batch_idx):
+        target_values = batch.pop("target_values")
+        target_weights = batch.pop("target_weights")
+        output_mask = batch.pop("output_mask")
+
+        # forward pass
+        output_values = self.model(**batch)
+
+        # compute loss
+        output_values = output_values[output_mask]
+        target_values = target_values[output_mask]
+        target_weights = target_weights[output_mask]
+
+        if self.modality_spec.loss_fn == "mse":
+            loss = torch.nn.functional.mse_loss(
+                output_values, target_values, reduction="none"
+            )
+            loss = loss * target_weights[:, None]
+            loss = loss.mean()
+        else:
+            raise NotImplementedError("Only MSE loss is supported for now.")
```

Comment:
why are we going backwards here?

Reply:
Just wanted to get it done without having to test _this_ code on datasets with other kinds of losses. We can bring back other losses when we've tested them. Imo we shouldn't include a feature we haven't tested. Adding more losses can be separate PRs.

Reply:
we have `compute_loss_or_metric` that we can use no?

Reply:
forgot about compute_loss_or_metric. fixed.